baremetal/metal3machine_manager.go

```diff
@@ -1338,15 +1335,58 @@ func (m *MachineManager) SetNodeProviderID(ctx context.Context, bmhID, providerI
 		m.Log.Info("Target node is not found, requeuing")
 		return &RequeueAfterError{RequeueAfter: requeueAfter}
 	}
+	namespace := m.Metal3Machine.GetNamespace()
+	m3mName := m.Metal3Machine.GetName()
+	bmhName, err := m.getBmhNameFromM3Machine()
+	if err != nil {
+		m.Log.Info("unable to retrieve BMH from Metal3Cluster")
+		return &RequeueAfterError{RequeueAfter: requeueAfter}
+	}
+	providerIDLegacy := *providerID
+	providerIDNew := fmt.Sprintf("metal3://%s/%s/%s", namespace, bmhName, m3mName)
+	if !m.Metal3Cluster.Spec.NoCloudProvider {
+		// Settig providerID on nodes takes time. Continuous monitoring is required.
+		nodesWithProviderIDset := 0
+		for _, node := range nodes.Items {
+			m.Log.Info("PP0: ", node.Spec.ProviderID)
+			if node.Spec.ProviderID == "" {
+				m.Log.Info("PP1: NoCloudProvider set to false. But, no providerID is set on the node either", node.GetName())
+				nodesWithProviderIDset++
+			} else if node.Spec.ProviderID == providerIDNew {
+				m.Log.Info("PP2: Found a node with matching node providerID in new format")
+				*providerID = providerIDNew
+				return nil
+			} else if node.Spec.ProviderID == providerIDLegacy {
+				m.Log.Info("PP3: Found a node with matching node providerID in legacy format")
+				*providerID = providerIDLegacy
+				return nil
+			}
```

Comment:
missing a case here, where the PID is set but to something different (all other provisioned nodes)

Reply:
thanks, done